### ISSUE TITLE
Limit reading from ZIP files

### ISSUE THREAD

**Issue by thrust26** [stella-emu/stella#791]
People sometimes have huge ZIP files. And when Stella finds one, it tries to read them, effectively freezing for a very long time.

We should limit the maximum files size here.

**Comment by thrust26** [comment]
@sa666666 Can you take care of this? I wouldn't know where to start.

**Comment by sa666666** [comment]
Yes, I have some ideas, and I'm working on that part of the code now anyway.

What should the limits be?  And should they only be for ZIP files, or all files in general?  And I guess we have to take MVC into account too.  For those, there is no technical limit, since we're streaming instead of loading the entire thing into a buffer.

**Comment by thrust26** [comment]
I suppose the size is not the problem, but the number of files contained. What do you think? Should we make some tests?

**Comment by sa666666** [comment]
Yeah, I guess we need to find out exactly _why_ it's slowing down on ZIP file before we can decide how to fix the issue.

**Comment by thrust26** [comment]
A 6.5 GB file containing 19 AVIs freezes Stella. So size matters.

**Comment by thrust26** [comment]
How about adding a progress bar instead? Like the one used in launcher when sub-directories are searched too.

**Comment by thrust26** [comment]
I tested another file, this time 10GB (all AVIs). This resulted into a stack overflow in `ZipHandler::next()`. A 2GB file works without problem. I suppose a ZIP size larger than 2GB is not handled by our code.

**Comment by sa666666** [comment]
There are some places (commented) in the ZIP code that are still using 32-bit int instead of 64-bit.  I meant to come back to them.  If this is all the issue is, then it's probably not too hard to fix.  But I still question if we even need to handle 256MB ZIP files.  That could contain every ROM ever created a dozen times over.

**Comment by thrust26** [comment]
The progress bar for each file list can be enabled by commenting line 159 of `FileListWidget`. It will not show immediately, only after 2500 files have been processed. So it works well if we process a large number of files. 

Not sure about a low number of large files, I suppose these are processed fast?

**Comment by sa666666** [comment]
I really need to make up some test cases to see exactly why it's failing.

**Comment by Keatah** [comment]
Maybe simply read zip files with a certain string or codeword in their filename. Stella would only try to read zips with "4Stella" in the name. Everything else is ignored. Example:

MyRomzCollection4Stella.zip
..is parsed and read.

Astronomy.zip
Setup.zip
screenshotutil.zip
..are totally ignored.

**Comment by andrew-davie** [comment]
Is the issue reading the large file
A quick look at zip file formats suggests to me first you'd look at the central directory - which is at the end. I'd be reading small blocks from the end of the file until I found the directory id # 0x02014b50 and then parsing that. After that, I'd be stepping through the individual entries using the length info to step between them. If you have the ability to random-access to the file, then you'd be able to avoid reading all the large file data.  I'm not sure if I'm understanding the problem correctly?

**Comment by skitt** [comment]
@andrew-davie that is what the existing code does, as far as I understand it: https://github.com/stella-emu/stella/blob/master/src/common/ZipHandler.cxx (look for `myCdPos`).

**Comment by thrust26** [comment]
Correct. The problem comes from huge zip files, often > 1GB.

**Comment by thrust26** [comment]
If we remove line 160 in `FileListWidget::incProgress()` this might mitigate the problem. This would cause a progress bar display with cancel option for all long running operations.

**Comment by sa666666** [comment]
Current progress:
- files are only treated as ROMs and loaded if they have a valid file extension
- MVC ROMs are detected, and loaded using the first 8K only

This fixes several of the above issues.  Still TODO the following:
- check ROM size and compare to `Cart::maxSize()`; this requires work to `FilesystemNode` to detect file size
- extend ZIP support to 64-bit; may not be necessary, as I don't know if we need to care about ZIP files > 2GB

**Comment by sa666666** [comment]
Everything is now done for this issue, except for support 64-bit ZIP files, which really belongs as its own issue.  Commit is in b5cc7d41c9987717e36e3e977c2bea96e2d8800d.
